include/util/multi_level_partition.hpp

```diff
@@ -1,31 +1,310 @@
+        return bit_to_level[msb];
+    }
+
+    std::size_t GetNumberOfLevels() const final override { return level_offsets.size(); }
+
+    std::size_t GetNumberOfCells(LevelID level) const final override
+    {
+        return GetCell(level, GetSenitileNode());
+    }
+
+    // Returns the lowest cell id (at `level - 1`) of all children `cell` at `level`
+    CellID BeginChildren(LevelID level, CellID cell) const final override
+    {
+        BOOST_ASSERT(level > 1);
+        auto lidx = LevelIDToIndex(level);
+        auto offset = level_to_children_offset[lidx];
+        return cell_to_children[offset + cell];
+    }
+
+    // Returns the highest cell id (at `level - 1`) of all children `cell` at `level`
+    CellID EndChildren(LevelID level, CellID cell) const final override
+    {
+        BOOST_ASSERT(level > 1);
+        auto lidx = LevelIDToIndex(level);
+        auto offset = level_to_children_offset[lidx];
+        return cell_to_children[offset + cell + 1];
+    }
+
+  private:
+    inline std::size_t LevelIDToIndex(LevelID l) const { return l - 1; }
+
+    // We save the sentinel as last node in the partition information.
+    // It has the highest cell id in each level so we can derived the range
+    // of cell ids efficiently.
+    inline NodeID GetSenitileNode() const { return partition.size() - 1; }
+
+    void SetCellID(LevelID l, NodeID node, std::size_t cell_id)
+    {
+        auto lidx = LevelIDToIndex(l);
+
+        auto shifted_id = cell_id << level_offsets[lidx];
+        auto cleared_cell = partition[node] & ~level_masks[lidx];
+        partition[node] = cleared_cell | shifted_id;
+    }
+
+    // If we have N cells per level we need log_2 bits for every cell ID
+    std::vector<std::size_t>
+    makeLevelOffsets(const std::vector<std::size_t> &level_to_num_cells) const
+    {
+        std::vector<std::size_t> offsets;
+        offsets.reserve(level_to_num_cells.size());
+
+        auto sum_bits = 0;
+        for (auto num_cells : level_to_num_cells)
+        {
+            // bits needed to number all contained vertexes
+            auto bits = static_cast<std::uint64_t>(std::ceil(std::log2(num_cells + 1)));
+            offsets.push_back(sum_bits);
+            sum_bits += bits;
+            BOOST_ASSERT(sum_bits < 64);
+        }
+        // sentinel
+        offsets.push_back(sum_bits);
+
+        return offsets;
+    }
+
+    std::vector<PartitionID> makeLevelMasks() const
+    {
+        std::vector<PartitionID> masks;
+        masks.reserve(level_offsets.size());
+
+        util::for_each_pair(level_offsets.begin(),
+                            level_offsets.end(),
+                            [&](const auto offset, const auto next_offset) {
+                                // create mask that has `bits` ones at its LSBs.
+                                // 000011
+                                BOOST_ASSERT(offset < sizeof(PartitionID) * 8);
```

Comment:
It can be removed if `makeLevelOffsets` guarantees `offset <= NUM_PARTITION_BITS`